rules-tests/test.js

```diff
@@ -0,0 +1,165 @@
+    let db = authedApp({ uid: TEST_UID });
+    let usrProfile = db.collection('users').doc(TEST_UID);
+    await usrProfile.set({ lastName: 'Smith' });
+    db = authedApp({ uid: TEST_UID2 });
+    usrProfile = db.collection('users').doc(TEST_UID);
+    await firebase.assertFails(usrProfile.get());
+  });
+  it("doesn't allow setting of the role", async () => {
+    const db = authedApp({ uid: TEST_UID });
+    const usrProfile = db.collection('users').doc(TEST_UID);
+    await firebase.assertFails(usrProfile.set({ role: 'system-admin' }));
+  });
+
+  // users_privileged
+  it('prevents un-authenticated user from creating profile', async () => {
+    const db = authedApp(null);
+    const usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await firebase.assertFails(usrProfile.set({ lastName: 'Smith' }));
+  });
+  it('allows authenticated user to create profile', async () => {
+    const db = authedApp({ uid: TEST_UID });
+    const usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await firebase.assertSucceeds(usrProfile.set({ lastName: 'Smith' }));
+  });
+  it("doesn't allow a different user to create profile", async () => {
+    const db = authedApp({ uid: TEST_UID2 });
+    const usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await firebase.assertFails(usrProfile.set({ lastName: 'Smith' }));
+  });
+  it('prevents un-authenticated user from reading profile', async () => {
+    let db = authedApp({ uid: TEST_UID });
+    let usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await usrProfile.set({ lastName: 'Smith' });
+    db = authedApp(null);
+    usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await firebase.assertFails(usrProfile.get());
+  });
+  it('allows authenticated user to read profile', async () => {
+    let db = authedApp({ uid: TEST_UID });
+    let usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await usrProfile.set({ lastName: 'Smith' });
+    db = authedApp({ uid: TEST_UID });
+    usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await firebase.assertSucceeds(usrProfile.get());
+  });
+  it("doesn't allow another user to read profile", async () => {
+    let db = authedApp({ uid: TEST_UID });
+    let usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await usrProfile.set({ lastName: 'Smith' });
+    db = authedApp({ uid: TEST_UID2 });
+    usrProfile = db.collection('users_privileged').doc(TEST_UID);
+    await firebase.assertFails(usrProfile.get());
+  });
+
+  // users_public
+  it('prevents any creation of profiles', async () => {
+    const db = authedApp({ uid: TEST_UID });
+    const usrProfile = db.collection('users_public').doc(TEST_UID);
+    await firebase.assertFails(usrProfile.set({ d: { firstName: 'Tim' } }));
+  });
+  it('allows un-authenticated user to read profile', async () => {
+    let db = adminApp({ uid: TEST_UID });
```

Comment:
Is it better to initialize the admin once for the whole module?

Reply:
Not sure. Google's examples had it created for each test. Does it matter?

Reply:
I see.  It's probably not a big deal so we can ignore it for now.